### PR TITLE
Fixes to TwoBodyDecay dictionary and KFFitter/smoother nhits checks

### PR DESCRIPTION
**Pull request by usarica**
- Dictionary fix is needed for TwoBodyDecay constraint to produce the appropriate products when trackParameter constraint is used
- KFFitter/smoother nhits/n measurements=1 is possible when TBD constraint is applied. A cms::Exception is given if the following checks are not implemented.
- cms exception texts did not refer to the correct function, this is fixed as well.
- See relevant discussion after https://hypernews.cern.ch/HyperNews/CMS/get/recoTracking/1648/1/1/1/1/1/1/1.html

### CODE REVIEW

File: TrackingTools/TrajectoryState/src/classes_def.xml [hunk 1]
```diff
@@ -1,5 +1,6 @@
 <lcgdict>
   <!-- persistent="false" qualifier for non-wrappers is deprecated. Will be removed when no longer needed.-->
+  <class name="TrajectoryStateOnSurface"/>
```
Comment (VinInn) [inline TrackingTools/TrajectoryState/src/classes_def.xml]: add persistent="false" ?

Comment (usarica) [inline TrackingTools/TrajectoryState/src/classes_def.xml]: I was not quite sure what persistent does, and the comment just above said it is deprecated. Do you think I should still add it?

File: TrackingTools/TrackFitters/plugins/KFFittingSmoother.h [hunk 2]
```diff
@@ -118,7 +118,7 @@ namespace {
     Trajectory smoothingStep(Trajectory && fitted) const {
       if (theEstimateCut>0) {
 	// remove "outlier" at the end of Traj
-	while (!fitted.empty() &&
+	while (!fitted.empty() && fitted.foundHits()>=theMinNumberOfHits &&
```
Comment (VinInn) [inline TrackingTools/TrackFitters/plugins/KFFittingSmoother.h]: most probably "!fitted.empty()" is now redundant, anyhow cost is negligible...

File: TrackingTools/TrajectoryState/src/classes_def.xml (hunk 1 above)
Comment (VinInn) [inline TrackingTools/TrajectoryState/src/classes_def.xml]: ah ok, leave like this then... (sorry)

**Comment by VinInn** [comment]
+1  (should pass test, in principle no regression shall be observed)

**Comment by VinInn** [comment]
@cmsbuild , please test

**Comment by slava77** [comment]
it looks like this PR is superseded by #19583
please close this PR or expect to update #19583 after this gets merged.

**Comment by slava77** [comment]
@VinInn 
regarding your
> +1 (should pass test, in principle no regression shall be observed)
It seems like something is missing: Jenkins summary has ``` Reco comparison results: 8987 differences found in the comparisons```

e.g. in 136.731 (100 events on data)
- about 23% loss in electronGsfTracks
    - 10% less gedGsfElectrons 
- 10% less conversionStepTracks
- 3% less generalTracks

**Comment by VinInn** [comment]
excellent!
we have a new student that will have a look to gsf, this seems to be also an interesting case to study.
we will need full MTV etc.
no core resources in TRK-POG to devolve to this for the time being.
so -1

**Comment by usarica** [comment]
Is there a way to run a couple of the events run on these tests by ourselves? The contents of this PR are critical for the two other PRs rebased on top of this. If someone can point me to the files, I can probably find out what happens in gsf electron tracks.

**Comment by VinInn** [comment]
@usarica runTheMatrix etc

**Comment by usarica** [comment]
Thank you very much, this is very helpful. Let me check if I can write a small script to check which line changed causes the difference, and then see how to workaround it.

**Comment by usarica** [comment]
Ok, so I have some new results:
I ran the command @VinInn suggested and counted the tracks that succeed and fail. I produced two lists, one for the changes in KFTrajectoryFitter and one for KFTrajectorySmoother:

/afs/cern.ch/work/u/usarica/public/KFTS_KFTF/KFTF changes.txt
/afs/cern.ch/work/u/usarica/public/KFTS_KFTF/KFTS changes.txt

listing only those tracks that would have given a warning had I inserted LogError just before returning empty trajectories.

It looks like there are not many that fail the smoother (KFTS), but the primary cause is failures in the fitter (KFTF). I am now looking into a way to avoid this.

Do you think it is acceptable for me to commit the fix here first, so that checks can be re-run independent of the other commits in the PRs that reference this one, and then I port the change to those? Or would you suggest I commit directly to #19584 / #19583?

**Comment by slava77** [comment]
On 7/7/17 5:56 AM, Ulascan Sarica wrote:
> Do you think it is acceptable for me to commit the fix here first, so
> that checks can be re-run independent of the other commits in the PRs
> that reference this one, and then I port the change to those? Or would
> you suggest I commit directly to #19584
> <https://github.com/cms-sw/cmssw/pull/19584> / #19583
> <https://github.com/cms-sw/cmssw/pull/19583>?
> 


it's fine here

it will reset the signature and let me decide what to do

**Comment by usarica** [comment]
I updated the PR, reverting KFTrajectoryFitter and Smoother changes. The only code change remains in KFFittingSmoother.h, which is still needed for TBD constraint to work.

**Comment by slava77** [comment]
@cmsbuild please test

@usarica please clarify if this update removes all the differences or just most of them.
Thank you.

**Comment by usarica** [comment]
@slava77 I haven't noticed anything different by including or removing the change in KFFittingSmoother by itself. Differences seem to be solely due to KFTrajectoryFitter and KFTrajectorySmoother, which are reverted to the original in the latest commit.

**Comment by usarica** [comment]
It looks like the comparisons are said to be ready, the plots are missing.

**Comment by slava77** [comment]
-1

superseded by #19584

**Pull request closed without merging**